File: services/umbra/src/routing/intent-classifier.ts
import { OpenRouterClient } from '@umbra/shared';
import { Envelope, UmbraPayload, ModuleResult, SupportedLanguage } from '@umbra/shared';
import { Logger } from '@umbra/shared';

interface IntentClassificationResult {
  intent: string;
  confidence: number;
  targetModule: string;
  parameters?: Record<string, any>;
  requiresClarification: boolean;
}

export class IntentClassifier {
  private logger: Logger;
  private openRouterClient: OpenRouterClient;

  // Intent patterns for quick classification
  private intentPatterns = {
    finance_ocr: [
      /(?:extract|ocr|scan|read|process).*(?:document|invoice|receipt|bill|statement)/i,
      /(?:upload|send).*(?:invoice|receipt|bill|document)/i,
      /(?:analyze|parse).*(?:financial|invoice|receipt)/i
    ],
    finance_report: [
      /(?:generate|create|make).*(?:report|summary|analysis)/i,
      /(?:vat|tax|budget).*(?:report|summary)/i,
      /(?:financial|expense).*(?:report|analysis)/i
    ],
    vps_monitor: [
      /(?:check|show|display).*(?:status|health|system|server)/i,
      /(?:monitor|monitoring).*(?:system|server|vps)/i,
      /(?:cpu|memory|disk|uptime)/i
    ],
    vps_execute: [
      /(?:run|execute|start|stop|restart).*(?:command|script|service|container)/i,
      /(?:docker|container).*(?:start|stop|restart|remove)/i,
      /(?:deploy|update|manage).*(?:service|application)/i
    ],
    client_manage: [
      /(?:create|add|new).*(?:client|customer|account)/i,
      /(?:remove|delete).*(?:client|customer|account)/i,
      /(?:list|show).*(?:clients|customers|accounts)/i
    ],
    workflow_create: [
      /(?:create|build|make).*(?:workflow|automation|process)/i,
      /(?:n8n|workflow).*(?:create|build)/i,
      /(?:automate|automation)/i
    ],
    media_generate: [
      /(?:generate|create|make).*(?:image|video|audio|media)/i,
      /(?:create|generate).*(?:picture|photo|sound|music)/i,
      /(?:text.*speech|tts|voice)/i
    ],
    translation: [
      /(?:translate|translation).*(?:to|into)/i,
      /(?:convert|change).*(?:language)/i,
      /(?:french|english|portuguese|fr|en|pt)/i
    ],
    calculation: [
      /(?:calculate|compute|math|arithmetic)/i,
      /(?:\d+\s*[\+\-\*\/]\s*\d+)/,
      /(?:add|subtract|multiply|divide|sum)/i
    ]
  };

  constructor(openRouterClient: OpenRouterClient) {
    this.logger = new Logger('IntentClassifier');
    this.openRouterClient = openRouterClient;
  }

  async classifyIntent(envelope: Envelope<UmbraPayload>): Promise<ModuleResult<IntentClassificationResult>> {
    const startTime = Date.now();
    const { reqId, userId, lang, payload } = envelope;

    try {
      const message = payload.message;
      
      if (!message) {
        return {
          reqId,
          status: 'error',
          error: {
            type: 'functional',
            code: 'MISSING_MESSAGE',
            message: 'Message is required for intent classification',
            retryable: false
          }
        };
      }

      this.logger.debug('Classifying intent', {
        reqId,
        userId,
        messageLength: message.length,
        lang
      });

      // Try pattern-based classification first (faster)
      let patternResult = this.classifyByPatterns(message);
      
      // If pattern classification is uncertain, use AI
      if (!patternResult || patternResult.confidence < 0.7) {
        const aiResult = await this.classifyWithAI(message, lang);
        
        // Use AI result if it's more confident, otherwise combine
        if (aiResult.confidence > (patternResult?.confidence || 0)) {
          patternResult = aiResult;
        }
      }

      const durationMs = Date.now() - startTime;

      if (!patternResult || patternResult.confidence < 0.5) {
        return {
          reqId,
          status: 'needs_validation',
          data: {
            intent: 'clarification_needed',
            confidence: 0,
            targetModule: 'umbra',
            requiresClarification: true
          },
          audit: {
            module: 'umbra-classifier',
            durationMs
          }
        };
      }

      this.logger.audit('Intent classified', userId, {
        reqId,
        intent: patternResult.intent,
        confidence: patternResult.confidence,
        targetModule: patternResult.targetModule,
        method: patternResult.confidence >= 0.7 ? 'pattern' : 'ai'
      });

      return {
        reqId,
        status: 'success',
        data: patternResult,
        audit: {
          module: 'umbra-classifier',
          durationMs
        }
      };

    } catch (error) {
      const durationMs = Date.now() - startTime;
      
      this.logger.error('Intent classification failed', {
        reqId,
        userId,
        error: error.message
      });

      return {
        reqId,
        status: 'error',
        error: {
          type: 'technical',
          code: 'CLASSIFICATION_ERROR',
          message: 'Failed to classify user intent',
          retryable: true
        },
        audit: {
          module: 'umbra-classifier',
          durationMs
        }
      };
    }
  }

  private classifyByPatterns(message: string): IntentClassificationResult | null {
    const normalizedMessage = message.toLowerCase().trim();

    for (const [intent, patterns] of Object.entries(this.intentPatterns)) {
      for (const pattern of patterns) {
        if (pattern.test(normalizedMessage)) {
          const targetModule = this.getTargetModule(intent);
          
          return {
            intent,
            confidence: 0.8, // High confidence for pattern matches
            targetModule,
            requiresClarification: false
          };
        }
      }
    }

    return null;
  }

  private async classifyWithAI(message: string, lang: SupportedLanguage): Promise<IntentClassificationResult> {
    try {
      const result = await this.openRouterClient.classifyIntent(message, lang);
      
      return {
        intent: result.intent,
        confidence: result.confidence,
        targetModule: result.targetModule,
        requiresClarification: result.confidence < 0.7
      };
    } catch (error) {
      this.logger.warn('AI classification failed, using fallback', {
        error: error.message
      });

      return {
        intent: 'unknown',
        confidence: 0.1,
        targetModule: 'umbra',
        requiresClarification: true
      };
    }
  }

  private getTargetModule(intent: string): string {
    const moduleMapping: Record<string, string> = {
      finance_ocr: 'finance',
      finance_report: 'finance',
      vps_monitor: 'concierge',
      vps_execute: 'concierge',
      client_manage: 'business',
      workflow_create: 'production',
      media_generate: 'creator',
      translation: 'umbra',
      calculation: 'umbra'
    };

    return moduleMapping[intent] || 'umbra';
  }

  /**
   * Get clarification questions for ambiguous intents
   */
  getClarificationQuestions(message: string, lang: SupportedLanguage): string[] {
    const questions = {
      EN: [
        "I understand you want help, but could you be more specific?",
        "Are you looking to:",
        "• Process a financial document (invoice, receipt)?",
        "• Check system status or run commands?",
        "• Manage client accounts?",
        "• Create a workflow or automation?",
        "• Generate media (images, videos, audio)?",
        "• Translate text?",
        "• Perform calculations?",
        "",
        "Please let me know which of these options matches what you need."
      ],
      FR: [
        "Je comprends que vous voulez de l'aide, mais pourriez-vous être plus précis ?",
        "Cherchez-vous à :",
        "• Traiter un document financier (facture, reçu) ?",
        "• Vérifier l'état du système ou exécuter des commandes ?",
        "• Gérer des comptes clients ?",
        "• Créer un workflow ou une automatisation ?",
        "• Générer des médias (images, vidéos, audio) ?",
        "• Traduire du texte ?",
        "• Effectuer des calculs ?",
        "",
        "Merci de me faire savoir laquelle de ces options correspond à vos besoins."
      ],
      PT: [
        "Entendo que você quer ajuda, mas poderia ser mais específico?",
        "Você está procurando:",
        "• Processar um documento financeiro (fatura, recibo)?",
        "• Verificar status do sistema ou executar comandos?",
        "• Gerenciar contas de clientes?",
        "• Criar um fluxo de trabalho ou automação?",
        "• Gerar mídia (imagens, vídeos, áudio)?",
        "• Traduzir texto?",
        "• Realizar cálculos?",
        "",
        "Por favor, me informe qual dessas opções corresponde ao que você precisa."
      ]
    };

    return questions[lang] || questions.EN;
  }
}

File: python_services/umbra/umbra_service/routing/intent_classifier.py
"""Intent classification for Umbra requests."""

import re
from typing import Dict, List, Any, Optional
from datetime import datetime

from umbra_shared import (
    OpenRouterClient,
    UmbraLogger,
    Envelope,
    UmbraPayload,
    ModuleResult,
    SupportedLanguage,
)


class IntentClassificationResult:
    """Result of intent classification."""
    
    def __init__(
        self,
        intent: str,
        confidence: float,
        target_module: str,
        parameters: Optional[Dict[str, Any]] = None,
        requires_clarification: bool = False
    ):
        self.intent = intent
        self.confidence = confidence
        self.target_module = target_module
        self.parameters = parameters or {}
        self.requires_clarification = requires_clarification


class IntentClassifier:
    """Classify user intents for routing to appropriate modules."""
    
    def __init__(self, openrouter_client: OpenRouterClient):
        self.logger = UmbraLogger("IntentClassifier")
        self.openrouter_client = openrouter_client
        
        # Intent patterns for quick classification
        self.intent_patterns = {
            "finance_ocr": [
                r"(?:extract|ocr|scan|read|process).*(?:document|invoice|receipt|bill|statement)",
                r"(?:upload|send).*(?:invoice|receipt|bill|document)",
                r"(?:analyze|parse).*(?:financial|invoice|receipt)"
            ],
            "finance_report": [
                r"(?:generate|create|make).*(?:report|summary|analysis)",
                r"(?:vat|tax|budget).*(?:report|summary)",
                r"(?:financial|expense).*(?:report|analysis)"
            ],
            "vps_monitor": [
                r"(?:check|show|display).*(?:status|health|system|server)",
                r"(?:monitor|monitoring).*(?:system|server|vps)",
                r"(?:cpu|memory|disk|uptime)"
            ],
            "vps_execute": [
                r"(?:run|execute|start|stop|restart).*(?:command|script|service|container)",
                r"(?:docker|container).*(?:start|stop|restart|remove)",
                r"(?:deploy|update|manage).*(?:service|application)"
            ],
            "client_manage": [
                r"(?:create|add|new).*(?:client|customer|account)",
                r"(?:remove|delete).*(?:client|customer|account)",
                r"(?:list|show).*(?:clients|customers|accounts)"
            ],
            "workflow_create": [
                r"(?:create|build|make).*(?:workflow|automation|process)",
                r"(?:n8n|workflow).*(?:create|build)",
                r"(?:automate|automation)"
            ],
            "media_generate": [
                r"(?:generate|create|make).*(?:image|video|audio|media)",
                r"(?:ai|artificial intelligence).*(?:generate|create)",
                r"(?:text to|convert to).*(?:image|video|audio)"
            ],
            "simple_calculate": [
                r"(?:calculate|compute|math|arithmetic)",
                r"(?:\d+\s*[+\-*/]\s*\d+)",
                r"(?:what is|what's|whats).*(?:\d+.*\d+)"
            ],
            "simple_translate": [
                r"(?:translate|translation).*(?:to|into)",
                r"(?:french|english|portuguese).*(?:translation|translate)",
                r"(?:what does.*mean in|how do you say.*in)"
            ],
            "simple_help": [
                r"(?:help|assistance|support)",
                r"(?:what can you|what do you)",
                r"(?:how to|how do i|how can i)"
            ]
        }
        
        # Module mapping
        self.intent_to_module = {
            "finance_ocr": "finance",
            "finance_report": "finance",
            "vps_monitor": "concierge",
            "vps_execute": "concierge",
            "client_manage": "business",
            "workflow_create": "production",
            "media_generate": "creator",
            "simple_calculate": "umbra",
            "simple_translate": "umbra",
            "simple_help": "umbra"
        }
    
    async def classify_intent(
        self,
        envelope: Envelope[UmbraPayload]
    ) -> ModuleResult[IntentClassificationResult]:
        """Classify user intent from message."""
        start_time = datetime.utcnow()
        req_id = envelope.req_id
        
        try:
            message = envelope.payload.message or ""
            language = envelope.lang
            
            self.logger.debug("Classifying intent",
                            req_id=req_id,
                            message_length=len(message),
                            language=language)
            
            # First try pattern matching for fast classification
            pattern_result = self._classify_with_patterns(message)
            
            if pattern_result and pattern_result.confidence > 0.7:
                self.logger.info("Intent classified with patterns",
                               req_id=req_id,
                               intent=pattern_result.intent,
                               confidence=pattern_result.confidence)
                
                duration_ms = int((datetime.utcnow() - start_time).total_seconds() * 1000)
                
                return ModuleResult(
                    req_id=req_id,
                    status="success",
                    data=pattern_result,
                    audit={
                        "module": "umbra-intent-classifier",
                        "duration_ms": duration_ms,
                        "provider": "pattern_matching"
                    }
                )
            
            # Fall back to AI classification
            ai_result = await self._classify_with_ai(message, language)
            
            self.logger.info("Intent classified with AI",
                           req_id=req_id,
                           intent=ai_result.intent,
                           confidence=ai_result.confidence)
            
            duration_ms = int((datetime.utcnow() - start_time).total_seconds() * 1000)
            
            return ModuleResult(
                req_id=req_id,
                status="success",
                data=ai_result,
                audit={
                    "module": "umbra-intent-classifier",
                    "duration_ms": duration_ms,
                    "provider": "openrouter_ai"
                }
            )
            
        except Exception as e:
            duration_ms = int((datetime.utcnow() - start_time).total_seconds() * 1000)
            
            self.logger.error("Intent classification failed",
                            req_id=req_id,
                            error=str(e))
            
            return ModuleResult(
                req_id=req_id,
                status="error",
                error={
                    "type": "technical",
                    "code": "CLASSIFICATION_ERROR",
                    "message": f"Failed to classify intent: {str(e)}",
                    "retryable": True
                },
                audit={
                    "module": "umbra-intent-classifier",
                    "duration_ms": duration_ms
                }
            )
    
    def _classify_with_patterns(self, message: str) -> Optional[IntentClassificationResult]:
        """Classify intent using regex patterns."""
        message_lower = message.lower()
        best_match = None
        best_confidence = 0.0
        
        for intent, patterns in self.intent_patterns.items():
            for pattern in patterns:
                if re.search(pattern, message_lower, re.IGNORECASE):
                    # Calculate confidence based on pattern strength
                    confidence = 0.8  # Base confidence for pattern match
                    
                    # Boost confidence for exact keyword matches
                    if any(keyword in message_lower for keyword in ["invoice", "receipt", "document", "workflow"]):
                        confidence += 0.1
                    
                    if confidence > best_confidence:
                        best_confidence = confidence
                        best_match = intent
        
        if best_match:
            target_module = self.intent_to_module.get(best_match, "umbra")
            return IntentClassificationResult(
                intent=best_match,
                confidence=best_confidence,
                target_module=target_module,
                requires_clarification=best_confidence < 0.8
            )
        
        return None
    
    async def _classify_with_ai(
        self,
        message: str,
        language: SupportedLanguage
    ) -> IntentClassificationResult:
        """Classify intent using AI."""
        try:
            available_modules = ["finance", "concierge", "business", "production", "creator"]
            
            classification_result = await self.openrouter_client.classify_intent(
                message=message,
                language=language,
                available_modules=available_modules
            )
            
            intent = classification_result.get("intent", "unclear_request")
            target_module = classification_result.get("target_module", "umbra")
            confidence = float(classification_result.get("confidence", 0.5))
            
            return IntentClassificationResult(
                intent=intent,
                confidence=confidence,
                target_module=target_module,
                requires_clarification=confidence < 0.7
            )
            
        except Exception as e:
            self.logger.warning("AI classification failed, using fallback", error=str(e))
            
            return IntentClassificationResult(
                intent="unclear_request",
                confidence=0.1,
                target_module="umbra",
                requires_clarification=True
            )
    
    def get_clarification_prompts(
        self,
        intent: str,
        language: SupportedLanguage
    ) -> Dict[str, str]:
        """Get clarification prompts for unclear intents."""
        prompts = {
            "EN": {
                "finance_unclear": "I can help with financial documents. Are you trying to upload an invoice/receipt, or generate a financial report?",
                "vps_unclear": "I can help with VPS management. Do you want to check system status or execute a command?",
                "workflow_unclear": "I can help create workflows. What kind of automation are you looking to build?",
                "general_unclear": "I can help with finance, VPS management, business operations, workflows, or media generation. What would you like to do?"
            },
            "FR": {
                "finance_unclear": "Je peux vous aider avec les documents financiers. Essayez-vous de télécharger une facture/reçu, ou générer un rapport financier ?",
                "vps_unclear": "Je peux vous aider avec la gestion VPS. Voulez-vous vérifier l'état du système ou exécuter une commande ?",
                "workflow_unclear": "Je peux vous aider à créer des workflows. Quel type d'automatisation voulez-vous construire ?",
                "general_unclear": "Je peux vous aider avec les finances, la gestion VPS, les opérations commerciales, les workflows, ou la génération de médias. Que voulez-vous faire ?"
            },
            "PT": {
                "finance_unclear": "Posso ajudar com documentos financeiros. Você está tentando enviar uma fatura/recibo, ou gerar um relatório financeiro?",
                "vps_unclear": "Posso ajudar com gerenciamento VPS. Você quer verificar o status do sistema ou executar um comando?",
                "workflow_unclear": "Posso ajudar a criar workflows. Que tipo de automação você quer construir?",
                "general_unclear": "Posso ajudar com finanças, gerenciamento VPS, operações comerciais, workflows, ou geração de mídia. O que você gostaria de fazer?"
            }
        }
        
        return prompts.get(language, prompts["EN"])